### PR TITLE
[DS-4036] Delete EPersons even if they are referenced

### LINKED ISSUE #2803: DS-4036: Delete EPersons in DSpace
This is a duplicate of [DS-4036](https://jira.lyrasis.org/browse/DS-4036)  --- see this ticket for description

Estimate from release plan: 29 hours
Expressing interest: none

From the release plan sheet: REQUIRED FOR GDPR COMPLIANCE
Add test to verify that inprogress submission are deleted as well, including pending workflow tasks. Add test to verify that the EPerson is removed from Groups without touching other members. Add test to include item without submitter in Search and Browse endpoint. Add test to check edit of item with null submitter. Add tests related to verify that subscriptions and request-a-copy about such eperson are deleted as well. Add test to include submitter null item in the versioning items endpoint (or create a task for future check)

Add a configuration parameter to limit the DELETE operation to only administrator and implement proper security

### PR DESCRIPTION
**Pull request by pnbecker**
https://jira.duraspace.org/browse/DS-4036
Fixes https://github.com/DSpace/DSpace/issues/2803

DSpace references EPersons in different database tables like the
submitter of an item or like the EPerson that gets special rights
granted in the resourcepolicy table. This PR changes DSpace so it can
handle references that are set null instead of referencing an actual
EPerson. This is important to be able to delete EPersons which is
demanded by several data protection laws like GDPR in the European
Union.

### CODE REVIEW

**Comment by pnbecker** [comment]
I mark this as "work in progress". Not that I currently do have to add anything specific, but I think we will vet this and maybe work on this together as port of DSpace 7.

**Review by terrywbrady — APPROVED** [review]
Very nice work.  I have added a few questions and some minor suggestions to my review.

File: dspace-api/src/main/java/org/dspace/content/WorkspaceItemServiceImpl.java [hunk 1]
```diff
@@ -202,9 +202,8 @@ public void deleteAll(Context context, WorkspaceItem workspaceItem)
          */
         Item item = workspaceItem.getItem();
         if (!authorizeService.isAdmin(context)
-            && ((context.getCurrentUser() == null) || (context
-            .getCurrentUser().getID() != item.getSubmitter()
-                                             .getID()))) {
+            && (item.getSubmitter() == null || (context.getCurrentUser() == null)
```
[the remark this answers is not in the document]
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/content/WorkspaceItemServiceImpl.java]: When the session is not logged in.

File: dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/SingleUserReviewAction.java [hunk 2]
```diff
@@ -35,6 +35,7 @@
 
     public static final int MAIN_PAGE = 0;
     public static final int REJECT_PAGE = 1;
+    public static final int SUBMITTER_IS_DELETED_PAGE = 2;
```
[the remark this answers is not in the document]
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/SingleUserReviewAction.java]: A separate issue:  these states are defined identically in multiple classes.  Should they not be consolidated in a single place?

File: dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java [hunk 3]
```diff
@@ -5,31 +5,78 @@
...
+    protected WorkflowItemService workflowItemService = WorkflowServiceFactory.getInstance().getWorkflowItemService();
+    protected WorkflowService workflowService = WorkflowServiceFactory.getInstance().getWorkflowService();
+    protected WorkspaceItemService workspaceItemService = ContentServiceFactory.getInstance()
+                                                          .getWorkspaceItemService();
+
+    private Community community = null;
+    private Collection collection = null;
+    private Item item = null;
+
+    private static final String EMAIL = "kevin@dspace.org";
```
[the remark this answers is not in the document]
Comment (mwoodiupui) [reply dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java]: May I suggest using something like test@example.com, since that domain is reserved for testing.

File: dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java [hunk 4]
```diff
@@ -192,6 +202,15 @@ public void deleteByDsoEPersonPolicies(Context context, DSpaceObject dso, EPerso
 
     }
 
+    @Override
+    public void deleteAllEPersonPolicies(Context context, EPerson ePerson) throws SQLException {
+        String queryString = "delete from ResourcePolicy where eperson= :eperson";
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/authorize/dao/impl/ResourcePolicyDAOImpl.java]: It probably can. But in the same class there are a lot of comparable methods that look exactly the same. I did not change all of them to HQL as this would add code changes that have nothing to do with the deletion of EPersons. I would like to keep this PR as narrow as possible to reduce the risk of introducing new bugs. Shall we open up a separate ticket regarding the conversion of SQL to HQL?

File: dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java [hunk 5]
```diff
@@ -104,6 +104,17 @@ protected ItemDAOImpl() {
         return iterate(query);
     }
 
+    @Override
+    public Iterator<Item> findBySubmitter(Context context, EPerson eperson, boolean retrieveAllItems)
+        throws SQLException {
+        Query query = createQuery(context, "FROM Item WHERE submitter= :submitter");
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/content/dao/impl/ItemDAOImpl.java]: This extends an existing method, that was written in SQL. As mentioned above, we did not change all occurrences of SQL to HQL to keep the PR narrow.

File: dspace-api/src/main/java/org/dspace/eperson/EPersonServiceImpl.java [hunk 6]
```diff
@@ -179,45 +211,192 @@ public EPerson create(Context context) throws SQLException, AuthorizeException {
...
+                    "You must be an admin to delete an EPerson");
         }
-
         // check for presence of eperson in tables that
         // have constraints on eperson_id
         List<String> constraintList = getDeleteConstraints(context, ePerson);
-
-        // if eperson exists in tables that have constraints
-        // on eperson, throw an exception
         if (constraintList.size() > 0) {
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/eperson/EPersonServiceImpl.java]: I understand the idea. But if an EPerson is referenced in some tables, we want to ask the administrator if he/she is sure that the EPerson should be deleted (for example here: https://github.com/DSpace/DSpace/pull/2229/files#diff-7f92e562d6c3ce16aca99e10e5abf821R265). The idea is to have one method that is used to list all occurrences in the code that tests if we need to ask questions to the administrator as well as in the deletion method. Like this we can avoid different ways to check if an EPerson is referenced in a specific table or not.
For example we ask an administrator if an EPerson really should be deleted if Workspace Items would be deleted together with the EPerson.

File: dspace-api/src/main/java/org/dspace/workflowbasic/BasicWorkflowServiceImpl.java [hunk 7]
```diff
@@ -798,33 +798,38 @@ protected void notifyOfArchive(Context context, Item item, Collection coll)
...
-
-            // Get title
-            String title = item.getName();
-            if (StringUtils.isBlank(title)) {
-                try {
-                    title = I18nUtil.getMessage("org.dspace.workflow.WorkflowManager.untitled");
-                } catch (MissingResourceException e) {
-                    title = "Untitled";
+            // send the notification only if the person was not deleted in the
+            // meantime between submission and archiving.
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/workflowbasic/BasicWorkflowServiceImpl.java]: Will be changed in our next commit. Thank you!

File: dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/AcceptEditRejectAction.java [hunk 8]
```diff
@@ -34,6 +34,7 @@
 
     public static final int MAIN_PAGE = 0;
     public static final int REJECT_PAGE = 1;
+    public static final int SUBMITTER_IS_DELETED_PAGE = 2;
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/AcceptEditRejectAction.java]: Good catch! I changed `AcceptEditRejectAction.proccessMainPage(...)`. No it will be returned if the reject button is pressed and the submitter is set null. Will be part of our next commit.

File: dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/ReviewAction.java [hunk 9]
```diff
@@ -33,6 +33,7 @@
 
     public static final int MAIN_PAGE = 0;
     public static final int REJECT_PAGE = 1;
+    public static final int SUBMITTER_IS_DELETED_PAGE = 2;
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/ReviewAction.java]: Good catch. I'll change `ReviewAction.processMainPage(...)`.

File: dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/SingleUserReviewAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/SingleUserReviewAction.java]: @terrywbrady Will be solved as above.
@mwoodiupui I would agree. But I would like to keep it separated of this ticket. I see enough potential to introduce new bugs in this PR. So Everything that can be done in a separate PR/Issue should be done there. Do you think it is worth a ticket?

File: dspace-api/src/main/resources/Messages.properties [hunk 10]
```diff
@@ -1824,6 +1838,11 @@ In response to your request I have the pleasure to send you in attachment a copy
 Best regards,\n\
 {3} <{4}>
 
+itemRequest.admin.response.body.approve = Dear {0},\n\
+In response to your request I have the pleasure to send you in attachment a copy of the file(s) concerning the document: "{2}" ({1}).\n\n\
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/resources/Messages.properties]: Changed to "In response to your request I have the pleasure to send you in attachment a copy of the file(s) concerning the document:" (remove "a" in front of copy).

File: dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java]: I expect that we copied that from some other test, but I agree. Changed following @mwoodiupui proposal. The email address was also hardcoded in the `destroy()` method. I referenced the static field there.

File: dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java [hunk 11]
```diff
@@ -71,687 +120,416 @@ public void destroy() {
...
+        try {
+            ePersonService.delete(context, deleteEperson);
+        } catch (AuthorizeException | IOException ex) {
+            log.error("Cannot delete EPersion, caught " + ex.getClass().getName() + ":", ex);
+            fail("Caught an Exception while deleting an EPerson. " + ex.getClass().getName() +
+                 ": " + ex.getMessage());
+        }
+        context.restoreAuthSystemState();
+        context.commit();
+        EPerson fundDeletedEperson = ePersonService.findByEmail(context, EMAIL);
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/test/java/org/dspace/eperson/EPersonTest.java]: Renamed to `findDeletedEpersion`.

**Comment by pnbecker** [comment]
@terrywbrady @mwoodiupui Thank you for the review. Either I answered directly to your questions/suggestions or I changed the PR in the PR I just pushed. Can you please take a look if my answers/changes are satisfying?

File: dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/SingleUserReviewAction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mwoodiupui) [reply dspace-api/src/main/java/org/dspace/xmlworkflow/state/actions/processingaction/SingleUserReviewAction.java]: You might just leave a comment there and not make any code changes at this time.  I only noted it as a potential future problem.

File: dspace-api/src/main/java/org/dspace/authorize/service/AuthorizeService.java [hunk 12]
```diff
@@ -417,6 +417,16 @@ public void removePoliciesActionFilter(Context context, DSpaceObject dso, int ac
...
 
+    /**
+     * Removes all policies from an eperson that belong to an EPerson.
+     *
+     * @param c current context
+     * @param e the eperson
+     * @throws SQLException if there's a database problem
+     * @throws AuthorizeException if authorization error
+     */
+    public void removeAllEPersonPolicies(Context c, EPerson e) throws SQLException, AuthorizeException;
```
[the remark this answers is not in the document]
Comment (pnbecker) [reply dspace-api/src/main/java/org/dspace/authorize/service/AuthorizeService.java]: @terrywbrady You're probably right. But we have this mismatch also between all of the other methods of the `ResourcePolicyService` and the `ResourcePolicyDAO`. In the DAO the term used is always "remove", in the Service it is "delete".

**Comment by pnbecker** [comment]
I rebased the PR and resolved the conflicts.

File: dspace-api/src/main/java/org/dspace/eperson/EPersonServiceImpl.java [hunk 13]
```diff
@@ -179,45 +211,192 @@ public EPerson create(Context context) throws SQLException, AuthorizeException {
...
+                        List<XmlWorkflowItem> xmlWorkflowItems = xmlWorkflowItemService
+                                                                 .findBySubmitter(context, ePerson);
+
+                        for (XmlWorkflowItem xmlWorkflowItem : xmlWorkflowItems) {
+                            ClaimedTask pooledTask = claimedTaskService
+                                                     .findByWorkflowIdAndEPerson(context, xmlWorkflowItem, ePerson);
+                            xmlWorkflowService.deleteClaimedTask(context, xmlWorkflowItem, pooledTask);
+
+                            try {
+                                workflowRequirementsService.removeClaimedUser(context, xmlWorkflowItem,
```
Comment (benbosman) [inline dspace-api/src/main/java/org/dspace/eperson/EPersonServiceImpl.java]: What happens if the user is in the pool, will they be reassigned?
What happens if the user is/was the only user in the pool?

File: dspace-api/src/main/java/org/dspace/eperson/EPersonServiceImpl.java [hunk 14]
```diff
@@ -179,45 +211,192 @@ public EPerson create(Context context) throws SQLException, AuthorizeException {
...
+                        // we delete the EPerson, it won't need any rights anymore.
+                        authorizeService.removeAllEPersonPolicies(context, ePerson);
+                    } else if (StringUtils.equals(tableName, "tasklistitem") && isBasicFramework) {
+                        // remove EPerson from the list of EPersons that may claim some specific workflow tasks.
+                        TaskListItemService taskListItemService = BasicWorkflowServiceFactory.getInstance()
+                                                                  .getTaskListItemService();
+                        taskListItemService.deleteByEPerson(context, ePerson);
+                    } else if (StringUtils.equals(tableName, "cwf_pooltask") && isXmlFramework) {
+                        PoolTaskService poolTaskService = XmlWorkflowServiceFactory.getInstance().getPoolTaskService();
+                        poolTaskService.deleteByEperson(context, ePerson);
```
Comment (benbosman) [inline dspace-api/src/main/java/org/dspace/eperson/EPersonServiceImpl.java]: what happens if they're the only person in the pool

**Review by benbosman — CHANGES_REQUESTED** [review]
The behavior when the user is part of the workflow is not clear to me yet, it seems the item would move to different workflow state when deleting a user. This is something the repository admin may not expect to happen. I did also add a few inline comments specific to the workflow.
It's very hard to verify whether all use cases are covered where it's currently assumed an item has a submitter. So this implies if some parts are missing, a NullPointerException would still pop up.

If the user's confidential data (name, email address, …) would be replaced instead of deleting the EPerson, all these changes would no longer be required. That would avoid all the changes here, eliminates the risk of forgetting any customizations, and would ensure the items remain in their current state.
An alternative solution would be to link all the items to a single EPerson account, which will have the same advantages, but would require that account to be not deletable.

**Comment by pnbecker** [comment]
Ben,

thank you for the review. I missed it in December and got aware of it by the end of last week. I know about the possibility to link "orphaned items" to a single user, often called ghost user. I agree, my approach is risking NPEs. But I'm doing this for the sake of a cleaner implementation in the long run. I want to avoid having to care about a ghost user everywhere, where we are dealing with EPersons.

You pointed out onw side-effects that we missed in deed. Thank you for finding that issue. A user being deleted may be the last user in a Workflow group. I think we have two bugs in DSpace that explains how I would expect DSpace to handle that case:

1. If the group of users for a specific Workflow step is empty this step is skipped silently. If an administrator activates a workflow step, but leaves the group empty, this resolves in an unexpeted behavior: having a workflow step that is activated but skipped nevertheless. I think DSpace should keep the item in the workflow step, send a mail to the administrator and put a warning into the log files.

2. If a user is added to a workflow step group, workflow item tasks are not assigned to that user. To make this concrete: if you have a workflow step with 500 items in the task pool and add a user to that workflow step, the user is still not seeing a single workflow item, until a new one is put into the pool. Then the user sees the new task only but not any of the tasks that existed before he/she was added to that workflow step's group.

If we would resolve both bugs, DSpace would not skip any steps anymore and administrators would have a chance to give people access to workflow tasks whose workflow groups are empty. This might also be a way to solve https://jira.duraspace.org/browse/DS-1941.

I know that we want to move to the XMLWorkflowSteps in DSpace 7. All my notes above relate to the default Workflows of DSpace 6 and I have to test if these issues are the same for the XMLWorkflows.

**Comment by benbosman** [comment]
@pnbecker 

* If the group of users for a specific Workflow step is empty, this step is skipped silently in the XML Workflow as well. So this will require a solution to avoid the workflow step from being skipped (e.g. refusing to delete an EPerson if it would create an empty workflow group, so the admin would need to solve that problem first)
* If a user is added to a workflow step group, workflow item tasks which are in the pool are assigned to that user in the XML Workflow, so that problem should be solved because the Basic Workflow was dropped.
